{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ded165bb-e6eb-4832-a98b-b995678ba43d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import re\n",
    "import os\n",
    "import math\n",
    "import gzip\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fc3d0032-f4eb-48d4-a9de-d8fa81a37b57",
   "metadata": {},
   "outputs": [],
   "source": [
    "def loadFasta(fasta_path, as_dict=False,uppercase=False, stop_at=None,\n",
    "              revcomp=False):\n",
    "    fastas = []\n",
    "    seq = None\n",
    "    header = None\n",
    "    for r in (gzip.open(fasta_path) if fasta_path.endswith(\".gz\") else open(fasta_path)):\n",
    "        if type(r) is bytes: r = r.decode(\"utf-8\")\n",
    "        r = r.strip()\n",
    "        if r.startswith(\">\"):\n",
    "            if seq != None and header != None:\n",
    "                fastas.append([header, seq])\n",
    "                if stop_at != None and len(fastas) >= stop_at:\n",
    "                    break\n",
    "            seq = \"\"\n",
    "            header = r[1:]\n",
    "        else:\n",
    "            if seq != None:\n",
    "                seq += r.upper() if uppercase else r\n",
    "            else:\n",
    "                seq = r.upper() if uppercase else r\n",
    "    if stop_at != None and len(fastas) < stop_at:\n",
    "        fastas.append([header, seq])\n",
    "    elif stop_at == None:\n",
    "        fastas.append([header, seq])\n",
    "    if as_dict:\n",
    "        return {h: s for h, s in fastas}\n",
    "    if(revcomp):\n",
    "        for rec in fastas:\n",
    "            rc=generate_complementary_sequence(rec[1])\n",
    "            rec[1]=rec[1]+\"NNNNNNNNNNNNNNNNNNNN\"+rc\n",
    "    return pd.DataFrame({'location': [e[0].split('=')[-1] for e in fastas],\n",
    "                         'sequence': [e[1] for e in fastas]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a0318982-060d-46d3-8cf8-f8fff0b4cf10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert GR and AP1 to low activation sequences at varying distances\n",
    "# output format is for prediction\n",
    "forward_ap1 = 'TGAGTCAT'\n",
    "forward_gr  = 'GAACATTATGTTC'\n",
    "\n",
    "def insert_motifs_pred(df, seq_len=300):\n",
    "    seq = df['sequence']\n",
    "    loc = df['location']\n",
    "    ls_distance = []\n",
    "    ls_ref = []\n",
    "    ls_allele = []\n",
    "    ls_seq = []\n",
    "\n",
    "    for i in range(len(df)):\n",
    "        sequence = seq[i]\n",
    "        max_dist = seq_len - len(forward_gr)\n",
    "        dist = np.random.randint(len(forward_ap1), max_dist + 1)  # dist: from begining of a motif to the begining of another motif\n",
    "        \n",
    "        if dist == max_dist: pos_ap1 = 0\n",
    "        else: pos_ap1 = np.random.randint(0, seq_len - len(forward_gr) - dist)\n",
    "        pos_gr = pos_ap1 + dist\n",
    "\n",
    "        sequence = sequence[:pos_ap1] + forward_ap1 + sequence[pos_ap1 + len(forward_ap1):]\n",
    "        sequence = sequence[:pos_gr] + forward_gr + sequence[pos_gr + len(forward_gr):]\n",
    "        \n",
    "        ls_seq.append(sequence)\n",
    "        ls_distance.append(dist)\n",
    "        ls_ref.append('ref='+sequence[seq_len//2])\n",
    "        ls_allele.append(sequence[seq_len//2])\n",
    "        \n",
    "    installed = pd.DataFrame({'location':df['location'],'ref':ls_ref,'allele':ls_allele,'sequence':ls_seq})\n",
    "    loc_dist = pd.DataFrame({'location':df['location'],'distance':ls_distance})\n",
    "\n",
    "    return installed, loc_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "db35995b-e9b0-4d34-8091-fe66ab9c3ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_low_act_5000(in_dir, out_dir):\n",
    "    # get the 5000 lowest activation sequences from test set of Dex\n",
    "    # abs(median(log(Dex))-log(Dex)) close to 0\n",
    "    counts = pd.read_csv(in_dir+'/test-counts.txt.gz', compression=\"gzip\", sep='\\t', skiprows=1, header=None)\n",
    "    counts.columns = ['DNA1','DNA2','DNA3','DNA4','DNA5','RNA1','RNA2','RNA3','RNA4']\n",
    "    counts['DNA_aver'] = (counts['DNA1']+counts['DNA1']+counts['DNA1']+counts['DNA1']+counts['DNA1'])/5\n",
    "    counts['RNA_aver'] = (counts['RNA1']+counts['RNA2']+counts['RNA3']+counts['RNA4'])/4\n",
    "    counts['theta'] = np.log(counts['RNA_aver']/counts['DNA_aver'])\n",
    "    counts['activation'] = abs(counts['theta'].median()-counts['theta'])\n",
    "    counts.sort_values(by='activation',inplace=True)\n",
    "    counts = counts[:5000]\n",
    "    fasta = loadFasta(in_dir+'/test.fasta.gz')\n",
    "    fasta = fasta.loc[counts.index]\n",
    "    fasta.reset_index(drop=True, inplace=True)\n",
    "    fasta.to_csv(out_dir+'/low_act_5000.txt',sep='\\t',header=None,index=False)\n",
    "\n",
    "    # insert motifs at varying distances\n",
    "    installed, loc_dist = insert_motifs_pred(fasta)\n",
    "    installed.to_csv(out_dir+'/low_act_5000_installed.txt', sep='\\t',header=None,index=False)\n",
    "    loc_dist.to_csv(out_dir+'/low_act_5000_installed_dist.txt', sep='\\t',header=None,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6820e6ae-03af-49dc-84b5-a134b6535e07",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_low_act_5000('/hpc/group/igvf/A549/GR-AP1/simulated-seq/data/Dex-200', '/hpc/group/igvf/A549/GR-AP1/simulated-seq/data/Dex-200')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c482aaaa-7873-4864-a534-fa433bb442d1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c7e46ac-e182-47fe-8177-8db35426b7d0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7346afbc-abf5-4127-9426-69eb910b165f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNA1</th>\n",
       "      <th>DNA2</th>\n",
       "      <th>DNA3</th>\n",
       "      <th>DNA4</th>\n",
       "      <th>DNA5</th>\n",
       "      <th>RNA1</th>\n",
       "      <th>RNA2</th>\n",
       "      <th>RNA3</th>\n",
       "      <th>RNA4</th>\n",
       "      <th>DNA_aver</th>\n",
       "      <th>RNA_aver</th>\n",
       "      <th>theta</th>\n",
       "      <th>activation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1366065</th>\n",
       "      <td>0.998980</td>\n",
       "      <td>1.064236</td>\n",
       "      <td>1.050659</td>\n",
       "      <td>0.978319</td>\n",
       "      <td>1.056019</td>\n",
       "      <td>1.040806</td>\n",
       "      <td>3.053206</td>\n",
       "      <td>1.812080</td>\n",
       "      <td>0.894977</td>\n",
       "      <td>0.998980</td>\n",
       "      <td>1.700267</td>\n",
       "      <td>0.531806</td>\n",
       "      <td>1.519404e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447648</th>\n",
       "      <td>1.382367</td>\n",
       "      <td>1.356350</td>\n",
       "      <td>1.190268</td>\n",
       "      <td>1.437359</td>\n",
       "      <td>1.470051</td>\n",
       "      <td>2.671450</td>\n",
       "      <td>2.024191</td>\n",
       "      <td>2.982936</td>\n",
       "      <td>1.732599</td>\n",
       "      <td>1.382367</td>\n",
       "      <td>2.352794</td>\n",
       "      <td>0.531806</td>\n",
       "      <td>1.519404e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715169</th>\n",
       "      <td>1.168348</td>\n",
       "      <td>1.247945</td>\n",
       "      <td>1.234613</td>\n",
       "      <td>1.197423</td>\n",
       "      <td>1.137119</td>\n",
       "      <td>1.930261</td>\n",
       "      <td>1.363602</td>\n",
       "      <td>1.965881</td>\n",
       "      <td>2.694383</td>\n",
       "      <td>1.168348</td>\n",
       "      <td>1.988532</td>\n",
       "      <td>0.531806</td>\n",
       "      <td>2.630957e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93294</th>\n",
       "      <td>1.481037</td>\n",
       "      <td>1.545088</td>\n",
       "      <td>1.491052</td>\n",
       "      <td>1.604272</td>\n",
       "      <td>1.503496</td>\n",
       "      <td>2.305634</td>\n",
       "      <td>3.074418</td>\n",
       "      <td>2.758016</td>\n",
       "      <td>1.944841</td>\n",
       "      <td>1.481037</td>\n",
       "      <td>2.520727</td>\n",
       "      <td>0.531805</td>\n",
       "      <td>1.217074e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530535</th>\n",
       "      <td>1.261267</td>\n",
       "      <td>1.122617</td>\n",
       "      <td>1.420087</td>\n",
       "      <td>1.392818</td>\n",
       "      <td>1.342155</td>\n",
       "      <td>2.762899</td>\n",
       "      <td>2.507008</td>\n",
       "      <td>1.795506</td>\n",
       "      <td>1.521303</td>\n",
       "      <td>1.261267</td>\n",
       "      <td>2.146679</td>\n",
       "      <td>0.531805</td>\n",
       "      <td>1.351904e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1049252</th>\n",
       "      <td>1.339384</td>\n",
       "      <td>1.452326</td>\n",
       "      <td>1.471403</td>\n",
       "      <td>1.308361</td>\n",
       "      <td>1.280040</td>\n",
       "      <td>2.051518</td>\n",
       "      <td>3.038055</td>\n",
       "      <td>1.405221</td>\n",
       "      <td>2.640495</td>\n",
       "      <td>1.339384</td>\n",
       "      <td>2.283822</td>\n",
       "      <td>0.533640</td>\n",
       "      <td>1.834194e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>773396</th>\n",
       "      <td>1.186309</td>\n",
       "      <td>1.224786</td>\n",
       "      <td>1.158240</td>\n",
       "      <td>1.339675</td>\n",
       "      <td>1.150456</td>\n",
       "      <td>2.263488</td>\n",
       "      <td>1.130021</td>\n",
       "      <td>2.752427</td>\n",
       "      <td>1.945314</td>\n",
       "      <td>1.186309</td>\n",
       "      <td>2.022812</td>\n",
       "      <td>0.533642</td>\n",
       "      <td>1.835532e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354826</th>\n",
       "      <td>8.507863</td>\n",
       "      <td>8.518713</td>\n",
       "      <td>8.679517</td>\n",
       "      <td>9.032385</td>\n",
       "      <td>9.263957</td>\n",
       "      <td>24.750328</td>\n",
       "      <td>12.423927</td>\n",
       "      <td>9.887226</td>\n",
       "      <td>10.966610</td>\n",
       "      <td>8.507863</td>\n",
       "      <td>14.507023</td>\n",
       "      <td>0.533642</td>\n",
       "      <td>1.835852e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1628904</th>\n",
       "      <td>1.075799</td>\n",
       "      <td>1.220727</td>\n",
       "      <td>1.211549</td>\n",
       "      <td>1.146359</td>\n",
       "      <td>1.067875</td>\n",
       "      <td>1.113617</td>\n",
       "      <td>1.707532</td>\n",
       "      <td>3.033037</td>\n",
       "      <td>1.483329</td>\n",
       "      <td>1.075799</td>\n",
       "      <td>1.834379</td>\n",
       "      <td>0.533642</td>\n",
       "      <td>1.835966e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>672815</th>\n",
       "      <td>2.549307</td>\n",
       "      <td>2.917673</td>\n",
       "      <td>2.679267</td>\n",
       "      <td>2.556236</td>\n",
       "      <td>2.648679</td>\n",
       "      <td>4.456667</td>\n",
       "      <td>1.945909</td>\n",
       "      <td>6.202355</td>\n",
       "      <td>4.718951</td>\n",
       "      <td>2.549307</td>\n",
       "      <td>4.330970</td>\n",
       "      <td>0.529970</td>\n",
       "      <td>1.836220e-03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             DNA1      DNA2      DNA3      DNA4      DNA5       RNA1  \\\n",
       "1366065  0.998980  1.064236  1.050659  0.978319  1.056019   1.040806   \n",
       "447648   1.382367  1.356350  1.190268  1.437359  1.470051   2.671450   \n",
       "715169   1.168348  1.247945  1.234613  1.197423  1.137119   1.930261   \n",
       "93294    1.481037  1.545088  1.491052  1.604272  1.503496   2.305634   \n",
       "530535   1.261267  1.122617  1.420087  1.392818  1.342155   2.762899   \n",
       "...           ...       ...       ...       ...       ...        ...   \n",
       "1049252  1.339384  1.452326  1.471403  1.308361  1.280040   2.051518   \n",
       "773396   1.186309  1.224786  1.158240  1.339675  1.150456   2.263488   \n",
       "354826   8.507863  8.518713  8.679517  9.032385  9.263957  24.750328   \n",
       "1628904  1.075799  1.220727  1.211549  1.146359  1.067875   1.113617   \n",
       "672815   2.549307  2.917673  2.679267  2.556236  2.648679   4.456667   \n",
       "\n",
       "              RNA2      RNA3       RNA4  DNA_aver   RNA_aver     theta  \\\n",
       "1366065   3.053206  1.812080   0.894977  0.998980   1.700267  0.531806   \n",
       "447648    2.024191  2.982936   1.732599  1.382367   2.352794  0.531806   \n",
       "715169    1.363602  1.965881   2.694383  1.168348   1.988532  0.531806   \n",
       "93294     3.074418  2.758016   1.944841  1.481037   2.520727  0.531805   \n",
       "530535    2.507008  1.795506   1.521303  1.261267   2.146679  0.531805   \n",
       "...            ...       ...        ...       ...        ...       ...   \n",
       "1049252   3.038055  1.405221   2.640495  1.339384   2.283822  0.533640   \n",
       "773396    1.130021  2.752427   1.945314  1.186309   2.022812  0.533642   \n",
       "354826   12.423927  9.887226  10.966610  8.507863  14.507023  0.533642   \n",
       "1628904   1.707532  3.033037   1.483329  1.075799   1.834379  0.533642   \n",
       "672815    1.945909  6.202355   4.718951  2.549307   4.330970  0.529970   \n",
       "\n",
       "           activation  \n",
       "1366065  1.519404e-07  \n",
       "447648   1.519404e-07  \n",
       "715169   2.630957e-07  \n",
       "93294    1.217074e-06  \n",
       "530535   1.351904e-06  \n",
       "...               ...  \n",
       "1049252  1.834194e-03  \n",
       "773396   1.835532e-03  \n",
       "354826   1.835852e-03  \n",
       "1628904  1.835966e-03  \n",
       "672815   1.836220e-03  \n",
       "\n",
       "[5000 rows x 13 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_train = loadFasta('/datacommons/igvf-pm/A549/full-set/Dex-200/300-bases/data-normalized/train.fasta.gz')\n",
    "# get 5000 low-activation sequence from training set of Dex: abs(median(log(Dex))-log(Dex)) close to 0 from experimental counts\n",
    "counts = pd.read_csv('/datacommons/igvf-pm/A549/full-set/Dex-200/300-bases/data-normalized/train-counts.txt.gz', compression=\"gzip\", sep='\\t', skiprows=1, header=None)\n",
    "counts.columns = ['DNA1','DNA2','DNA3','DNA4','DNA5','RNA1','RNA2','RNA3','RNA4']\n",
    "counts['DNA_aver'] = (counts['DNA1']+counts['DNA1']+counts['DNA1']+counts['DNA1']+counts['DNA1'])/5\n",
    "counts['RNA_aver'] = (counts['RNA1']+counts['RNA2']+counts['RNA3']+counts['RNA4'])/4\n",
    "counts['theta'] = np.log(counts['RNA_aver']/counts['DNA_aver'])\n",
    "counts['activation'] = abs(counts['theta'].median()-counts['theta'])\n",
    "counts.sort_values(by='activation',inplace=True)\n",
    "counts = counts[:5000]\n",
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "735fbae5-3140-49e1-b23b-d338bce35622",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>location</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1366065</th>\n",
       "      <td>chr3:71940000-71940300</td>\n",
       "      <td>CTACTGAGAGGCCCTTTAGCTAATTGAAAGGAGAAAGTGGTTCCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447648</th>\n",
       "      <td>chr15:98665350-98665650</td>\n",
       "      <td>TGCATCTGCCTCCGTTTGAAGCATCCTCCTCCAGTAAGTGGGTCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715169</th>\n",
       "      <td>chr10:2105300-2105600</td>\n",
       "      <td>TGTTCCGGATGGTGAATCCTCACGAGTCTGCAGCAACCTCAATTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93294</th>\n",
       "      <td>chr19:10350700-10351000</td>\n",
       "      <td>AGACAGGAGTAAGGCACACGGTGTGGGTGAGGCTGACATCCCCCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530535</th>\n",
       "      <td>chr13:99978000-99978300</td>\n",
       "      <td>ATGAACACTGGCCCCCGCGGCCCTGAGCTGGATCCGCTCAAGCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1049252</th>\n",
       "      <td>chr7:55952550-55952850</td>\n",
       "      <td>ATGGTGAAACTCTGTCTTTACTAAAAATACAAAAATTAGCTCGGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>773396</th>\n",
       "      <td>chr10:102655350-102655650</td>\n",
       "      <td>GGGCCATTTCCAAATTGAGATCTGCAGTTTCCTGCACCTGTAGCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354826</th>\n",
       "      <td>chr16:46390650-46390950</td>\n",
       "      <td>AATCATTATCGAATGGAATCGAATGGAATCATCGAATGGAATTGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1628904</th>\n",
       "      <td>chr1:91737800-91738100</td>\n",
       "      <td>AGACGGTGTCGGGACAAGATTCAAGGTCTCCAGGTATGGTCTGAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>672815</th>\n",
       "      <td>chr11:72578750-72579050</td>\n",
       "      <td>TTTGTCACCCTGAGATGGTCTAGGTTTCTGTCTCCCCAGAACACAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          location  \\\n",
       "1366065     chr3:71940000-71940300   \n",
       "447648     chr15:98665350-98665650   \n",
       "715169       chr10:2105300-2105600   \n",
       "93294      chr19:10350700-10351000   \n",
       "530535     chr13:99978000-99978300   \n",
       "...                            ...   \n",
       "1049252     chr7:55952550-55952850   \n",
       "773396   chr10:102655350-102655650   \n",
       "354826     chr16:46390650-46390950   \n",
       "1628904     chr1:91737800-91738100   \n",
       "672815     chr11:72578750-72579050   \n",
       "\n",
       "                                                  sequence  \n",
       "1366065  CTACTGAGAGGCCCTTTAGCTAATTGAAAGGAGAAAGTGGTTCCAT...  \n",
       "447648   TGCATCTGCCTCCGTTTGAAGCATCCTCCTCCAGTAAGTGGGTCAT...  \n",
       "715169   TGTTCCGGATGGTGAATCCTCACGAGTCTGCAGCAACCTCAATTCT...  \n",
       "93294    AGACAGGAGTAAGGCACACGGTGTGGGTGAGGCTGACATCCCCCTC...  \n",
       "530535   ATGAACACTGGCCCCCGCGGCCCTGAGCTGGATCCGCTCAAGCCCC...  \n",
       "...                                                    ...  \n",
       "1049252  ATGGTGAAACTCTGTCTTTACTAAAAATACAAAAATTAGCTCGGCG...  \n",
       "773396   GGGCCATTTCCAAATTGAGATCTGCAGTTTCCTGCACCTGTAGCTC...  \n",
       "354826   AATCATTATCGAATGGAATCGAATGGAATCATCGAATGGAATTGAT...  \n",
       "1628904  AGACGGTGTCGGGACAAGATTCAAGGTCTCCAGGTATGGTCTGAGA...  \n",
       "672815   TTTGTCACCCTGAGATGGTCTAGGTTTCTGTCTCCCCAGAACACAG...  \n",
       "\n",
       "[5000 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "low_act_seq = all_train.loc[counts.index]\n",
    "low_act_seq.to_csv('/datacommons/igvf-pm/A549/GR-AP1/simulated-seq/data/Dex-200/diluting/low_act_5000.txt', sep='\\t', header=None, index=False)\n",
    "low_act_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "420b7e89-d20e-49e5-b00a-aeed71410a64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>location</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr22:10716750-10717050</td>\n",
       "      <td>TAACATGTCACGTACTATAAGGTTTGCATGGCATGGTGGCAGCTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr22:10716850-10717150</td>\n",
       "      <td>TTTTTCCTTAAACATTTGAATCCATTCAATGTCATTCCATTCCAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr22:10716950-10717250</td>\n",
       "      <td>TCAATCCCATTAAATTCTACTCCATTCCATTCTTTTCCATTCCATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr22:10717000-10717300</td>\n",
       "      <td>TCCATTTGAACCCAATGAATTCCACTGCATTCAATTGAAGTCCATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22:10717050-10717350</td>\n",
       "      <td>TCCATTCGGGTCCATTCCCTTCAATTCCATTCGTGTCAATTCCATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700565</th>\n",
       "      <td>chr1:248934500-248934800</td>\n",
       "      <td>GGGGAATAGAAGAAAGAAAGAAAAGGAAGTTGCCCAGGGATAGTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700566</th>\n",
       "      <td>chr1:248934650-248934950</td>\n",
       "      <td>ACACACACACACACACACACACACACACACACACATATAAAAATAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700567</th>\n",
       "      <td>chr1:248934700-248935000</td>\n",
       "      <td>TAGTTGTGACTTTATAATCTTTGAGGAAGAACTTTCCTCAAAGTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700568</th>\n",
       "      <td>chr1:248934750-248935050</td>\n",
       "      <td>CAGTGCTTTGTAAGCATTGTCTCCATAAAAGTCAACCTTACTTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700569</th>\n",
       "      <td>chr1:248934950-248935250</td>\n",
       "      <td>ATAAGAGTATTCAGTAAGGATTTCAAAATTCTGGAAAAATCAGGCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1695570 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         location  \\\n",
       "0         chr22:10716750-10717050   \n",
       "1         chr22:10716850-10717150   \n",
       "2         chr22:10716950-10717250   \n",
       "3         chr22:10717000-10717300   \n",
       "4         chr22:10717050-10717350   \n",
       "...                           ...   \n",
       "1700565  chr1:248934500-248934800   \n",
       "1700566  chr1:248934650-248934950   \n",
       "1700567  chr1:248934700-248935000   \n",
       "1700568  chr1:248934750-248935050   \n",
       "1700569  chr1:248934950-248935250   \n",
       "\n",
       "                                                  sequence  \n",
       "0        TAACATGTCACGTACTATAAGGTTTGCATGGCATGGTGGCAGCTGA...  \n",
       "1        TTTTTCCTTAAACATTTGAATCCATTCAATGTCATTCCATTCCAAT...  \n",
       "2        TCAATCCCATTAAATTCTACTCCATTCCATTCTTTTCCATTCCATT...  \n",
       "3        TCCATTTGAACCCAATGAATTCCACTGCATTCAATTGAAGTCCATT...  \n",
       "4        TCCATTCGGGTCCATTCCCTTCAATTCCATTCGTGTCAATTCCATT...  \n",
       "...                                                    ...  \n",
       "1700565  GGGGAATAGAAGAAAGAAAGAAAAGGAAGTTGCCCAGGGATAGTTA...  \n",
       "1700566  ACACACACACACACACACACACACACACACACACATATAAAAATAG...  \n",
       "1700567  TAGTTGTGACTTTATAATCTTTGAGGAAGAACTTTCCTCAAAGTTT...  \n",
       "1700568  CAGTGCTTTGTAAGCATTGTCTCCATAAAAGTCAACCTTACTTCCT...  \n",
       "1700569  ATAAGAGTATTCAGTAAGGATTTCAAAATTCTGGAAAAATCAGGCA...  \n",
       "\n",
       "[1695570 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# exclude 5000 low-activation sequences from whole training set to select remaining sequences\n",
    "rdm_seq = all_train.loc[~all_train.index.isin(counts.index)]\n",
    "rdm_seq.to_csv('/datacommons/igvf-pm/A549/GR-AP1/simulated-seq/data/Dex-200/diluting/rdm_seq.txt', sep='\\t', header=None, index=False)\n",
    "rdm_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "927f5a1b-1979-4510-8db0-608657100bf4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b1a084a-49ff-4452-8397-fb567de34b01",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "82b76116-8ffa-4b6e-8e1d-db6e5e8ed630",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get 5000 low-activation sequences for biased downsampling experiment, since train/test/validation were resplit\n",
    "test = loadFasta('/datacommons/igvf-pm/A549/GR-AP1/simulated-seq/data/Dex-200-biased/test.fasta.gz')\n",
    "all = loadFasta('/datacommons/igvf-pm/A549/processed-data/300-bases/Dex-200/Dex-200-all.fasta.gz')\n",
    "all_counts = pd.read_csv('/datacommons/igvf-pm/A549/processed-data/300-bases/Dex-200/Dex-200-all-normalized-counts.txt.gz', compression=\"gzip\", sep='\\t', skiprows=1, header=None)\n",
    "counts_idx = list(pd.merge(test, all, how='inner', on='location').index)\n",
    "counts = all_counts.loc[all_counts.index.isin(counts_idx)]\n",
    "\n",
    "counts = counts.reset_index(drop=True)\n",
    "counts.columns = ['DNA1','DNA2','DNA3','DNA4','DNA5','RNA1','RNA2','RNA3','RNA4']\n",
    "counts['DNA_aver'] = (counts['DNA1']+counts['DNA1']+counts['DNA1']+counts['DNA1']+counts['DNA1'])/5\n",
    "counts['RNA_aver'] = (counts['RNA1']+counts['RNA2']+counts['RNA3']+counts['RNA4'])/4\n",
    "counts['theta'] = np.log(counts['RNA_aver']/counts['DNA_aver'])\n",
    "counts['activation'] = abs(counts['theta'].median()-counts['theta'])\n",
    "counts = counts.sort_values(by='activation')\n",
    "counts = counts[:5000]\n",
    "\n",
    "test = test.loc[test.index.isin(counts.index)]\n",
    "test = test.reset_index(drop=True)\n",
    "\n",
    "installed, loc_dist = insert_motifs_pred(test)\n",
    "\n",
    "installed.to_csv('/datacommons/igvf-pm/A549/GR-AP1/simulated-seq/data/Dex-200-biased/low_act_5000_test.txt', sep='\\t', header=None, index=False)\n",
    "loc_dist.to_csv('/datacommons/igvf-pm/A549/GR-AP1/simulated-seq/data/Dex-200-biased/low_act_5000_test_dist.txt', sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "66b1d863-a54b-43b4-a1e4-309508fa74eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>location</th>\n",
       "      <th>ref</th>\n",
       "      <th>allele</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr22:10723400-10723700</td>\n",
       "      <td>ref=C</td>\n",
       "      <td>C</td>\n",
       "      <td>CATTCCATTCCATTCCATTAAATTCCATTCTGAGTCATTCCATTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr22:11435650-11435950</td>\n",
       "      <td>ref=T</td>\n",
       "      <td>T</td>\n",
       "      <td>GGTGGGGCTCCTGCCTGCCAATTTAGAAGGGGTGGGGCTCCCACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr22:17104800-17105100</td>\n",
       "      <td>ref=G</td>\n",
       "      <td>G</td>\n",
       "      <td>GACACTCCAGGTAGGGGACATGCGGCTGTCCTAGGCCATACTGGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr22:17591750-17592050</td>\n",
       "      <td>ref=C</td>\n",
       "      <td>C</td>\n",
       "      <td>ACCCGGGAGGCTGAGTCATCAGTGAGCCGAGATGGCACCACTGCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22:18131200-18131500</td>\n",
       "      <td>ref=A</td>\n",
       "      <td>A</td>\n",
       "      <td>GAGAACAGAACACTCTCCCCGCCCCAGCCTGATTCCTGCCTTACCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4995</th>\n",
       "      <td>chr1:242519700-242520000</td>\n",
       "      <td>ref=A</td>\n",
       "      <td>A</td>\n",
       "      <td>TGAGTCATGAGCAAGCAAGGGCCCCACTAGCCAAGAAGAGTCAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4996</th>\n",
       "      <td>chr1:242754400-242754700</td>\n",
       "      <td>ref=A</td>\n",
       "      <td>A</td>\n",
       "      <td>ATGTCAGAAAGTGCCTGGAAGTAGGGAGTGGTCTAAAGAATTTGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4997</th>\n",
       "      <td>chr1:242882150-242882450</td>\n",
       "      <td>ref=G</td>\n",
       "      <td>G</td>\n",
       "      <td>GCCTGCTCTGAGAAGACACCATGCAGCGTCTTCGTGTCTTGGGCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4998</th>\n",
       "      <td>chr1:245894050-245894350</td>\n",
       "      <td>ref=G</td>\n",
       "      <td>G</td>\n",
       "      <td>TCTGAGTCATTGAAGCCAGCTGGACTTCCAGGGTGGAGTGGGGAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4999</th>\n",
       "      <td>chr1:248934600-248934900</td>\n",
       "      <td>ref=C</td>\n",
       "      <td>C</td>\n",
       "      <td>TTCTGTCCAGGCATGCTGGAGCAAGCTAGGACAAGTGATTTGGAAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      location    ref allele  \\\n",
       "0      chr22:10723400-10723700  ref=C      C   \n",
       "1      chr22:11435650-11435950  ref=T      T   \n",
       "2      chr22:17104800-17105100  ref=G      G   \n",
       "3      chr22:17591750-17592050  ref=C      C   \n",
       "4      chr22:18131200-18131500  ref=A      A   \n",
       "...                        ...    ...    ...   \n",
       "4995  chr1:242519700-242520000  ref=A      A   \n",
       "4996  chr1:242754400-242754700  ref=A      A   \n",
       "4997  chr1:242882150-242882450  ref=G      G   \n",
       "4998  chr1:245894050-245894350  ref=G      G   \n",
       "4999  chr1:248934600-248934900  ref=C      C   \n",
       "\n",
       "                                               sequence  \n",
       "0     CATTCCATTCCATTCCATTAAATTCCATTCTGAGTCATTCCATTCC...  \n",
       "1     GGTGGGGCTCCTGCCTGCCAATTTAGAAGGGGTGGGGCTCCCACCT...  \n",
       "2     GACACTCCAGGTAGGGGACATGCGGCTGTCCTAGGCCATACTGGGA...  \n",
       "3     ACCCGGGAGGCTGAGTCATCAGTGAGCCGAGATGGCACCACTGCAC...  \n",
       "4     GAGAACAGAACACTCTCCCCGCCCCAGCCTGATTCCTGCCTTACCC...  \n",
       "...                                                 ...  \n",
       "4995  TGAGTCATGAGCAAGCAAGGGCCCCACTAGCCAAGAAGAGTCAGGC...  \n",
       "4996  ATGTCAGAAAGTGCCTGGAAGTAGGGAGTGGTCTAAAGAATTTGCT...  \n",
       "4997  GCCTGCTCTGAGAAGACACCATGCAGCGTCTTCGTGTCTTGGGCCA...  \n",
       "4998  TCTGAGTCATTGAAGCCAGCTGGACTTCCAGGGTGGAGTGGGGAAT...  \n",
       "4999  TTCTGTCCAGGCATGCTGGAGCAAGCTAGGACAAGTGATTTGGAAC...  \n",
       "\n",
       "[5000 rows x 4 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "installed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4e794869-532e-4b35-bad3-b091bfb12b67",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>location</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr22:10723400-10723700</td>\n",
       "      <td>201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr22:11435650-11435950</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr22:17104800-17105100</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr22:17591750-17592050</td>\n",
       "      <td>183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22:18131200-18131500</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4995</th>\n",
       "      <td>chr1:242519700-242520000</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4996</th>\n",
       "      <td>chr1:242754400-242754700</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4997</th>\n",
       "      <td>chr1:242882150-242882450</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4998</th>\n",
       "      <td>chr1:245894050-245894350</td>\n",
       "      <td>282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4999</th>\n",
       "      <td>chr1:248934600-248934900</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      location  distance\n",
       "0      chr22:10723400-10723700       201\n",
       "1      chr22:11435650-11435950        10\n",
       "2      chr22:17104800-17105100        28\n",
       "3      chr22:17591750-17592050       183\n",
       "4      chr22:18131200-18131500        95\n",
       "...                        ...       ...\n",
       "4995  chr1:242519700-242520000       272\n",
       "4996  chr1:242754400-242754700       149\n",
       "4997  chr1:242882150-242882450        37\n",
       "4998  chr1:245894050-245894350       282\n",
       "4999  chr1:248934600-248934900       117\n",
       "\n",
       "[5000 rows x 2 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loc_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0559ed27-5324-45d0-a141-e0b70403db32",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
